packages/contracts/evm-contracts/contracts/orderbook/OrderbookDex.sol

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+import {ERC165} from "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import {IERC165} from "@openzeppelin/contracts/utils/introspection/IERC165.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+import {IOrderbookDex} from "./IOrderbookDex.sol";
+
+/// @notice Facilitates trading an asset that is living on a different app-chain.
+contract OrderbookDex is IOrderbookDex, ERC165, ReentrancyGuard {
+    using Address for address payable;
+
+    mapping(uint256 => Order) public orders;
+    uint256 public ordersIndex;
+
+    error OrderIsInactive(uint256 orderId);
+    error Unauthorized();
+
+    /// @notice Returns the current index of orders (index that a new sell order will be mapped to).
+    function getOrdersIndex() public view returns (uint256) {
+        return ordersIndex;
+    }
+
+    /// @notice Returns the Order struct information about order of specified `orderId`.
+    function getOrder(uint256 orderId) public view returns (Order memory) {
+        return orders[orderId];
+    }
+
+    /// @notice Creates a sell order for the specified `assetAmount` at specified `price`.
+    /// @dev The order is saved in a mapping from incremental ID to Order struct.
+    function createSellOrder(uint256 assetAmount, uint256 price) public {
+        Order memory newOrder = Order({
+            seller: payable(msg.sender),
+            assetAmount: assetAmount,
+            price: price,
+            active: true
+        });
+        orders[ordersIndex] = newOrder;
+        emit OrderCreated(ordersIndex, msg.sender, assetAmount, price);
+        ordersIndex++;
```

Comment:
Although an increasing index works, it will make the UI slow to update (recall: we had the same problem in the inverse projection where if the inverse projection is base-initiated it takes a while for the game node to update)

In this case however, we can't easily rely on app-side initiation solely because new UTXO entries are created when an order is partially filled (which has a new index)

To solve it though, I think we can take a variation of what we did for the inverse projection system:

1. Initiate your sell order on Tarochi (by placing in your gold to trade). This will give you a wallet-specific counter (instead of a global counter like we have here)
2. The RPC endpoint we query to know is `https://${rpcBase}/inverseProjection/${purpose}/${chainIdentifier}/` just like in PRC3, but with `purpose` set to `orderbook`
3. Unlike the NFT case where we queried for the file `${address}/${userTokenId}`, we instead query for `${address}/${orderId}/${amount}`. The endpoint then returns something like `{ aboveInitial: true }` if the amount provided is above the initial tgold deposited by the user in (1)

Then, like PRC3, we can create a function in this contract (one with the default RPC, one with the user RPC) that constructs the RPC URL from the Solidity contract.

With this configuration, we should end up where the Solidity contract provides an endpoint that always works right away to know if the order is real (has enough tgold). You simply check
1. if `active === true` in the Solidity contract
2. if `aboveInitial === true` from the RPC call

Reply:
Note that this requires essentially 2 IDs that we need to work with:
1. An ID for the initial orderID
2. A sub-ID for how far we are in a chain of partial orders